### ISSUE TITLE
Conditional Types, implement custom error message for non matching case

### ISSUE THREAD

**Issue by abdulkareemnalband** [microsoft/TypeScript#33098]
## Search Terms
#conditionalType , #customErrorMessage

## Suggestion
In conditional types `never` is used to remove invalid types from assignments, Error message is is reported as **So and so type cannot be assigned to `never`** , Which  does not inform user what are the actual requirements for usecase.

Author of conditional type need to provide some custom error message for particular failure.
this could be provided as doc comment before `never`
Ex: 
```ts
type Test<C> = C extends some? C :/** Should extend something*/ never;
```

## Use Cases

Too improve the diagnostics of conditional types.

## Examples

```ts
type NonData<D> = D extends {data:any}?/**parameter should not contain property data */ never:D;

declare function nonData<T>(param:NonData<T>):void;


const withData = {data:10 , a:'a'};
nonData(withData); // error, but error message is "parameter should not contain property data "

const withNonData = { a:'a'};
nonData(withNonData); // pass
```

## Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by jack-williams** [comment]
Duplicate of #23689.

For now I tend to use something like:
```ts
type NonData<D> = D extends {data:any} ? [D, 'should not extend { data: any }', never] : D;
```
which is not fool-proof, but works well enough.

**Comment by abdulkareemnalband** [comment]
@jack-williams thanks
I could not find previous issues, 
closing this now in favor of #23689
